release.py

```diff
@@ -4,129 +4,98 @@
-            test.write("set(FSO_VERSION_MAJOR {})\n".format(self.version.major))
-            test.write("set(FSO_VERSION_MINOR {})\n".format(self.version.minor))
-            test.write("set(FSO_VERSION_BUILD {})\n".format(self.version.patch))
-            test.write("set(FSO_VERSION_REVISION 0)\n")
-            test.write("set(FSO_VERSION_REVISION_STR {})\n".format("-".join(self.version.prerelease)))
-
-    def allow_multiple_tags(self):
-        return True
-
-
+# Assumes git has SirKnightly credentials already
 def main():
-    script_state = ScriptState.load_from_file()
-
+    # Set up paths
+    print("Setting up paths...")
+    abspath = os.path.abspath(__file__)
+    dname = os.path.dirname(abspath)
+    os.chdir(dname)
+
+    # Parse the command line arguments
+    print("Parsing cmdline...")
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--config",     help="Sets the config file", default="config.yml")
+    parser.add_argument("--version",    help="The version to mark this release as")
+    parser.add_argument("--type",       help="Either \'candidate\' or \'release\'")
+    parser.add_argument("--candidate",  help="If --type = \'candidate\', this specifies the candidate number.  Is ignored if --type = \'release\'")
+
+    args = parser.parse_args()
+
+    # Read in configuration data from config.yml
+    print("Parsing config.yml...")
+    config = {}
+
+    with open(args.config, "r") as f:
+        try:
+            config = yaml.safe_load(f)
+            # Support some variables in the config
+            expand_config_vars(config)
+        except yaml.YAMLError as e:
+            print(e)
+            sys.exit(1)
+
+    # Checkout repo
+    print("Checking out repo: {}/{}".format(config["git"]["repo"], config["git"]["branch"]))
+    repo = git.GitRepository(config["git"]["repo"], config["git"]["branch"])
+    repo.update_repository()
+
+    # Verify version string is valid
+    print("Verifying version string...")
     if not semantic_version.validate(args.version):
-        print("Specified version is not a valid version string!")
-        return
+        print("Error: Specified version is not a valid version string!")
+        sys.exit(1)
 
+    # Form tag string according to type
+    print("Forming tag name...")
     version = semantic_version.Version(args.version)
-    if script_state is None:
-        # An existing script state overrides the commandline argument
-        if args.tag_name is not None:
-            script_state = ReleaseState(version)
-            script_state.state = ScriptState.STATE_TAG_PUSHED
-            script_state.tag_name = args.tag_name
-        else:
-            if script_state is None:
-                script_state = ReleaseState(version)
+    tag_name = ''
+    if (args.type == "candidate"):
+        # suffix RC# to version
+        if (not isinstance(args.candidate, int)) or (args.candidate < 0):
```

Comment:
This should be validated before we touch the repository above.

Reply:
`git.update_repository()` does a reset `--hard`, wouldn't moving the checkout to after the `args` validations delete `version_override.cmake`?

Reply:
To clarify, what I meant was to split the validation and execution so that this first validated the parameters and if they are valid, continue with the execution and update/reset the repository.